q2_types_genomics/per_sample_data/_format.py

```diff
@@ -0,0 +1,115 @@
+                            % (i, cells, self.EXPECTED_HEADER))
+                    else:
+                        header = cells
+                else:
+                    if len(cells) != len(header):
+                        raise ValidationError(
+                            'Line %d has %s cells (%s), expected %s.'
+                            % (i, len(cells), cells, len(header)))
+
+                    # Structure checks out, so let's make lookup easy
+                    cells = dict(zip(header, cells))
+
+                    # TODO: a bunch of tests in this subpackage aren't well
+                    # behaved --- many tests fail on this check because the
+                    # test data isn't constructed correctly. As well, there
+                    # appear to be framework-related issues preventing us from
+                    # making this kind of validation work for the relative
+                    # manifest formats at this time.
+                    if root == '':
+                        fp = os.path.join(root, cells[self.PATH_HEADER_LABEL])
+                        if not os.path.exists(os.path.expandvars(fp)):
+                            raise ValidationError(
+                                'File referenced on line %d could not be '
+                                'found (%s).'
+                                % (i, fp))
+
+                    records_seen += 1
+
+            if header is None:
+                raise ValidationError('No header found, expected: %s.'
+                                      % self.EXPECTED_HEADER)
+
+            if records_seen == 0:
+                raise ValidationError('No sample records found in manifest, '
+                                      'only observed comments, blank lines, '
+                                      'and/or a header row.')
+
+
+class MultiMAGManifestFormat(_FastaManifestBase):
+    EXPECTED_HEADER = ['sample-id', 'mag-id', 'filename']
+    PATH_HEADER_LABEL = 'filename'
+
+    def _validate_(self, level):
+        self._check_n_records(root=str(self.path.parent),
+                              n={'min': 10, 'max': None}[level])
+
+
+class MultiDirValidationMixin:
+    def _validate_(self, level):
+        for p in self.path.iterdir():
+            if not p.is_dir() and p.name not in ['MANIFEST', 'metadata.yml']:
+                raise ValidationError(
+                    "Files should be organised in per-sample directories")
+
+
+class MultiFASTADirectoryFormat(MultiDirValidationMixin,
+                                model.DirectoryFormat):
+    sequences = model.FileCollection(r'.+\.(fa|fasta)$', format=DNAFASTAFormat)
+
+    @sequences.set_path_maker
+    def sequences_path_maker(self, sample_id, mag_id):
+        # write out with fasta extension, regardless if input was fa or fasta
+        return '%s/%s.fasta' % (sample_id, mag_id)
+
+
+class MAGSequencesDirFmt(MultiFASTADirectoryFormat):
+    manifest = model.File('MANIFEST', format=MultiMAGManifestFormat)
```

Comment:
I'm sorry for being a little fuzzy on the specifics at this point, but is there still a need for `MAGSequencesDirFmt`?

It occurs to me you could create a dataframe property or method on the `MultiFASTADirectoryFormat`. And the semantics of it being a MAG over something else should be handled by the semantic type just fine.

Reply:
I think the main reason I did it this way was to be able to `qiime import` the sequences, where the `MultiFASTADirectoryFormat` -> `MAGSequencesDirFmt` is being used. Otherwise, I agree - this additional DirFmt shouldn't be necessary...